crates/revm/src/state_change.rs

```diff
@@ -46,6 +51,68 @@ pub fn post_block_balance_increments(
     balance_increments
 }
 
+/// Applies the pre-block call to the EIP-4788 beacon block root contract, using the given block,
+/// [ChainSpec], EVM.
+///
+/// If cancun is not activated or the block is the genesis block, then this is a no-op, and no
+/// state changes are made.
+#[inline]
+pub fn apply_beacon_root_contract_call<DB: Database>(
+    chain_spec: &ChainSpec,
+    block_timestamp: u64,
+    block_number: u64,
+    block_parent_beacon_block_root: Option<H256>,
+    evm: &mut EVM<State<DB>>,
+) -> Result<(), BlockExecutionError>
+where
+    <DB as Database>::Error: Debug,
+{
+    if chain_spec.is_cancun_activated_at_timestamp(block_timestamp) {
+        // if the block number is zero (genesis block) then the parent beacon block root must
+        // be 0x0 and no system transaction may occur as per EIP-4788
+        if block_number == 0 {
+            if block_parent_beacon_block_root != Some(H256::zero()) {
+                return Err(BlockValidationError::CancunGenesisParentBeaconBlockRootNotZero.into())
+            }
+        } else {
+            let parent_beacon_block_root = block_parent_beacon_block_root.ok_or(
+                BlockExecutionError::from(BlockValidationError::MissingParentBeaconBlockRoot),
+            )?;
+
+            // get previous env
+            let previous_env = evm.env.clone();
+
+            // modify env for pre block call
+            fill_tx_env_with_beacon_root_contract_call(&mut evm.env, parent_beacon_block_root);
+
+            let ResultAndState { mut state, .. } = match evm.transact() {
+                Ok(res) => res,
+                Err(e) => {
+                    evm.env = previous_env;
+                    return Err(BlockExecutionError::from(BlockValidationError::EVM {
+                        hash: Default::default(),
+                        message: format!("{e:?}"),
+                    }))
+                }
+            };
+
+            // TODO: re-evaluate this after 4788 implementers call, this is to prevent EIP161 state
+            // clearing of the system address, since it is touched during transact. The pyspec
+            // tests currently require the state to remain empty if it started empty, which
+            // somewhat contradicts EIP161, and clients still need consensus on what to do here.
+            state.remove(&SYSTEM_ADDRESS);
+            state.remove(&evm.env.block.coinbase);
```

Comment:
I think this current way makes sense? My default way of doing this EIP would've been w/o a smart contract but with direct state surgery, and that would involve no system address. So my intuition is that the system address should not be anywhere outside of this kind of weird functions.